server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.james.jmap.api.model.Upload;
+import org.apache.james.jmap.api.model.UploadId;
+import org.apache.james.jmap.api.model.UploadMetaData;
+import org.apache.james.jmap.api.upload.UploadRepository;
+import org.apache.james.mailbox.model.ContentType;
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class InMemoryUploadRepository implements UploadRepository {
+
+    private static final Map<UploadId, ImmutablePair<Username, BlobId>> uploadStore = new HashMap<>();
+    private static final Map<UploadId, UploadMetaData> uploadMetaDataStore = new HashMap<>();
+
+    static class UploadContext {
+        private final UploadId uploadId;
+        private final boolean exists;
+        private final byte[] data;
+
+        private UploadContext(UploadId uploadId,
+                              boolean exists,
+                              byte[] data) {
+            this.uploadId = uploadId;
+            this.exists = exists;
+            this.data = data;
+        }
+    }
+
+
+    private final BlobStore blobStore;
+    private final BucketName bucketName;
+
+    @Inject
+    public InMemoryUploadRepository(BlobStore blobStore) {
+        this.blobStore = blobStore;
+        this.bucketName = blobStore.getDefaultBucketName();
+    }
+
+    @Override
+    public Publisher<UploadId> upload(InputStream data, ContentType contentType, Username user) {
+        return Mono.just(checkUploadExits(data, user))
+            .flatMap(uploadContext -> {
+                if (!uploadContext.exists) {
+                    return store(uploadContext.data, uploadContext.uploadId, contentType, user);
+                } else {
+                    return Mono.just(uploadContext.uploadId);
+                }
+            });
+    }
+
+    @Override
+    public Publisher<Upload> retrieve(UploadId id, Username user) {
+        Preconditions.checkNotNull(id);
+        Preconditions.checkNotNull(user);
+
+        return retrieveBlobId(id, user)
+            .flatMap(blobId -> retrieveUpload(id, blobId));
+    }
+
+    private Mono<BlobId> retrieveBlobId(UploadId id, Username user) {
+        return Mono.justOrEmpty(uploadStore.get(id))
+            .filter(pair -> userHasAccessToUpload(user, pair.left))
+            .map(userAndBlobId -> userAndBlobId.right)
+            .switchIfEmpty(Mono.error(() -> new UploadNotFoundException(id)));
+    }
+
+    private Mono<Upload> retrieveUpload(UploadId id, BlobId blobId) {
+        return Mono.zip(Mono.from(blobStore.readBytes(bucketName, blobId)),
+            Mono.just(uploadMetaDataStore.get(id)))
+            .map(bytesAndMetadata -> Upload.from(bytesAndMetadata.getT2(), bytesAndMetadata.getT1()));
+    }
+
+    private boolean userHasAccessToUpload(Username userRequest, Username owner) {
+        return userRequest.equals(owner);
+    }
+
+    private UploadContext checkUploadExits(InputStream data, Username username) {
```

Comment:
Let's not care about deduplication at this level as the blobStore does take care of it for us.

Given that UploadId is randomly generated, we will never have conflicts. 

We can thus greatly simplify the code... (And get rid of `checkUploadExits`)

Reply:
is it in conflict with the wrote you did before? 
https://github.com/linagora/james-project/issues/4349#issuecomment-885495718

Reply:
MAY means that we are able to do it if we want, in no way that we have to do it ;-).

So no, no incompatibility at all ;-)

(And it is a quote of the JMAP spec serving another purpose there....)